### DISCUSSION TITLE
FMU & inputs

### DISCUSSION THREAD

**Discussion by ANaumann85** [OpenModelica/OpenModelica#9768] [Q&A] [answered]
Are input variables in modelica assumed to be constant in time? See for example the following simple model:

```mo
model testModel "test model"
  extends Modelica.Icons.Example;

  Real T;
  input Real Tref1 "reference Temperature";
  parameter Real alpha1 = 10;
  parameter Real T0 = 0 "start value for T";
  output Real q "output variable";
initial equation
  T = T0;
equation
   der(T) = alpha1*Tref1; 
   q = alpha1*(Tref1-T);
end testModel;
```

where Tref1 is assumed to be a time dependent input. 

When I generate an FMU in cosimulation mode with that model, I get completely wrong solutions. The root cause is, that the differentiates the equation for q under the assumption, that Tref is constant in time.

Is that expected behavior? How can I circumvent this assumption and generate an FMU, where the compiler does not apply this assumption?

**Comment by AnHeuermann** [comment, marked as answer]
> Are input variables in modelica assumed to be constant in time?

No, they are variable over time.
I could reproduce it and this looks like a bug to me.

**Reply by AnHeuermann** [reply]
> When I generate an FMU in cosimulation mode with that model, I get completely wrong solutions. 

I could reproduce your issue using OMSimulator:

```modelica
model TestModel "test model"
  Real T;
  input Real Tref1 "reference Temperature";
  parameter Real alpha1 = 10;
  parameter Real T0 = 0 "start value for T";
  output Real q "output variable";
initial equation
  T = T0;
equation
   der(T) = alpha1*Tref1; 
   q = alpha1*(Tref1-T);
end TestModel;
```

```modelica
model SomeInput
  output Real Tref1;
equation
  Tref1 = time^2;
end SomeInput;
```

```modelica
model connectedModel
  SomeInput someInput;
  TestModel testModel;
equation
  testModel.Tref1 = someInput.Tref1;
end connectedModel;
```
![image](https://user-images.githubusercontent.com/38031952/204253808-9fabe80f-2741-430f-978c-bac23136240a.png)

Now generate Co-Simulation FMUs for `TestModel` and `SomeInput` and connect them in an SSP model:
![image](https://user-images.githubusercontent.com/38031952/204254411-5fbbfd1b-09f9-4130-8b17-1ca08af03828.png)

The simulation result for `testModel.q` is constant at `zero` while `testModel.T` is correct.
![image](https://user-images.githubusercontent.com/38031952/204253697-81b7f3da-38cf-46cd-a88c-6c0ec34de084.png)

**Comment by ANaumann85** [comment]
Thank you for reproducing the problem. In the meantime I did some further experiments using this [repository](https://github.com/ANaumann85/openmodelica-fmu). 

In these tests, I compared the numerical solutions using FMU and modelica to analytical solutions.
According to the results, I think that the index reduction method is the culprit. Maybe together with some simplification.
